### PR TITLE
[WIP] Make classes for map/filter/zip

### LINKED ISSUE #503: Make filter/map/zip classes
Make `filter`/`map`/`zip` classes:
```Python
assert type(filter) is type
assert type(map) is type
assert type(zip) is type
```
Similarly to `enumerate` and `range`.

This is causing pickling errors in `test_builtin`.

### PR DESCRIPTION
**Pull request by yuhan0**
Resolves #503

This is just a direct copying of the map / filter / zip builtins from the Builtin.cs file into their own classes.

However it appears that the (CPython) 3 implementation of these classes differs from the current ones in more fundamental ways, returning stateful iterators instead of an enumerable.

``` python

x = map(lambda x:x+5, [10, 20 , 30])

print(list(x)) # => [15, 25, 35]
print(list(x)) # => []

```

Should these classes be rewritten to implement the `IEnumerator` interface instead of `IEnumerable`?

### CODE REVIEW

File: Src/IronPython/Runtime/Filter.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            IEnumerator e;
+            if(!PythonOps.TryGetEnumerator(context, iterable, out e)) {
+                throw PythonOps.TypeErrorForNotIterable(iterable);
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() {
+
+            if (_function != null && !PythonOps.IsCallable(_context, _function)) {
+                throw PythonOps.UncallableError(_function)
```
[the remark this answers is not in the document]
Comment (yuhan0) [reply Src/IronPython/Runtime/Filter.cs]: oops, forgot to push

**Comment by yuhan0** [comment]
Also, how do I run tests locally? Running this using mono on macOS, none of the build commands in Visual Studio work and the `make.ps1` command takes 1.5 minutes to run on each iteration. 
Is this expected? Still quite new to .NET development

**Comment by slozier** [comment]
You can't build IronPython.sln using your Visual Studio? Hmm, that's probably something that should be fixed, unfortunately I don't have a mac to try it.

Unfortunately I don't have much experience with building from the command line (I guess if I did I probably would have spent some time trying to optimizing it). One option would be to remove the automatic restore from `Build.proj` (remove `DependsOnTargets` in `<Target Name="Build" DependsOnTargets="RestoreReferences">`). I suspect most of the time is wasted building the test files.

For testing I usually run the test file directly during development (e.g. `mono bin/Debug/net45/ipy.exe Tests/test_regressions.py`. You can also run, for example `pwsh make.ps1 --frameworks net45 test-IronPython.test_regressions`.

**Comment by yuhan0** [comment]
Thanks! I was planning to make the IEnumerator change and add tests if necessary, but if you think that's for another PR then I'm fine with merging this one as it is :)

**Comment by slozier** [comment]
Thanks for the PR! I'll merge it in as-is and let you make another PR for the IEnumerator changes.

**Comment by yuhan0** [comment]
None of the tests seem to work:

```shell
$ mono bin/Release/net45/ipy.exe Tests/test_regressions.py
Traceback (most recent call last):
  File "Tests/test_regressions.py", line 17, in <module>
  File "/Users/yuhan/Github/ironpython3/bin/Release/net45/Lib/os.py", line 27, in <module>
ImportError: No module named stat
$ Traceback (most recent call last):
  File "Tests/test_assert.py", line 6, in <module>
ImportError: No module named unittest
```

Am I supposed to link it somehow to the system installation of Python where it looks for modules?

**Comment by slide** [comment]
You'll want to do something like:

```
IRONPYTHONPATH=/Users/yuhan/Github/ironpython3/Src/StdLib/Lib mono bin/Release/net45/ipy.exe Tests/test_regressions.py
```

Or just export the `IRONPYTHONPATH` variable.
